035-ca-abstraction.md

```diff
@@ -0,0 +1,315 @@
+In fact Strimzi has two such CAs.
+The "cluster CA" is used for issuing certificates for Strimzi components, such as ZooKeeper and Kafka brokers.
+The "clients CA" is used for issuing certificates for Kafka clients, for example via the User Operator.
+
+The cluster CA root certificate is added to Strimzi component trust stores (so that, for example, ZooKeeper nodes trust each other's and brokers' certificates, and brokers trust certs issued to other brokers, etc). 
+
+
+![Cluster CA relationships](./images/035-cluster-ca.svg)
+
+> In the diagram, the red lines show trust.
+> For example, because the Kafka Broker trusts the Cluster CA certificate, 
+> and the Cluster CA certificate signed the ZK EE certificate
+> the broker will trust the ZK EE certificate presented by the ZK node
+> during TLS handshake.
+
+The cluster CA root certificate also needs to be trusted by Kafka clients connecting to the cluster (so that clients trust the broker's EE certificates). 
+The client CA root certificate is added to the broker trust stores too, so that the brokers will trust certificates issued to end users. 
+
+![Client CA relationships](./images/035-clients-ca.svg)
+
+This provides the secure-by-default experience, but in addition to a buggy implementation (**TODO bug ref**) it has very signifiant limitations:
+
+1. It assumes that Strimzi is issuing certificates. 
+2. It doesn't really understand more complex certificate paths, i.e. intermediate certificates.
+3. It doesn't really understand that certificate validity is a property of the whole certificate path.
+4. It doesn't manage the removal of trust in old/unused certificates in a timely manner.
+
+## Motivation
+
+Removing the assumption that Strimzi always issues EE certificates using a CA certificate it has access to it would enable much more flexibility in CA handling.
+This will be compatible with commonly deployed certificate management solutions such as [Cert Manager][cmio], or [Vault][vault].
+This would be highly valuable for organisations with compliance requirements with regard to certificates.
+
+## Proposal
+
+It is assumed that we want Strimzi to:
+
+* continue to be secure by default. 
+* not have some mandatory dependency on some other project/product for certificate issuance. 
+
+It follows that Strimzi needs to be able to obtain EE certificates, but should not (necessarily) issue them directly itself.
+
+We therefore need an abstraction for how certificates can be issued. 
+In the default (backward compatible) mode this abtraction would continue to use `openssl`, but the new abstraction would also be compatible with issuance by external systems (or even, in theory, a human operator using tooling of their choice and interacting with well defined `Secrets`). 
+
+### Interface for certificate issuance
+
+```java
+public interface CertificateIssuer {
+
+    byte[] requestCertificate(Subject subject, int requestedValidity) throws CertificateIssuanceException;
```

Comment:
Not in this interface, no. The reason being is that different issuers (cert manager, vault) might not expose that. 
This could be handled in the API (see later) where we select which implementation is going to be used. For example with the `internal` type we could have 

```yaml
spec:
  clusterCa:
    validityDays: <integer>
    generateCertificateAuthority: <boolean>
    generateSecretOwnerReference: <boolean>
    renewalDays: <integer>
    certificateExpirationPolicy: <renew-certificate|replace-key>
    keySize: 4096
    keyType: RSA # (or EC)
```

The `internal` type could also support an implementation (e.g. `openssl` vs `bouncycastle`), but it's not part of this proposal.

Reply:
Does the validity really differ from key size, type etc.? Different providers might have limited availabilities of that as well I guess.

Reply:
Good point, maybe this should be configured away too. I can't see a reason why we need it in the interface for the cluster CA (it's not like we want ZK certs to have different validity to broker certs for example). For the clients CA I suppose it is possible that some `KafkaUser` could support an explicit validity period (possibly even end dates) which would allow time-limited accounts. Generally speaking there are better ways of achieving that with Kafka than doing it at the certificate level.